contracts/ts/genMaciState.ts

```diff
@@ -55,38 +59,81 @@ const genMaciStateFromContract = async (
 
     assert(stateTreeDepth === maciState.stateTreeDepth)
 
-    // Fetch event logs
-    const initLogs = await provider.getLogs({
-        ...maciContract.filters.Init(),
-        fromBlock: fromBlock,
-    })
+    let initLogs: any[] = []
+    let signUpLogs: any[] = []
+    let mergeStateAqSubRootsLogs: any[] = []
+    let mergeStateAqLogs: any[] = []
+    let deployPollLogs: any[] = []
+    
+    const lastBlock = endBlock ? endBlock : await provider.getBlockNumber()
+    
+    // Fetch event logs in batches
+    for (let i = fromBlock; i < lastBlock; i += blocksPerRequest + 1) {
+        const toBlock = (i + blocksPerRequest) >= lastBlock ? undefined : i + blocksPerRequest
+
+        const tmpInitLogs = await provider.getLogs({
+            ...maciContract.filters.Init(),
+            fromBlock: i,
+            toBlock,
+            address: address
+        })
+
+        // sleep to avoid rate limiting
+        if (sleepAmount) await sleep(sleepAmount)
+
+        initLogs = initLogs.concat(tmpInitLogs)
+
+        const tmpSignUpLogs = await provider.getLogs({
+            ...maciContract.filters.SignUp(),
+            fromBlock: i,
+            toBlock,
+            address: address
+        })
+
+        if (sleepAmount) await sleep(sleepAmount)
+
+        signUpLogs = signUpLogs.concat(tmpSignUpLogs)
+    
+        const tmpMergeStateAqSubRootsLogs = await provider.getLogs({
+            ...maciContract.filters.MergeStateAqSubRoots(),
+            fromBlock: i,
+            toBlock,
+            address: address
+        })
```

Comment:
Since we're querying the maci contract for multiple events, it might be more performant if we call getLogs() with multiple event topics, (e.g. don't need to sleep in between each batch of events). 

https://docs.ethers.org/v5/concepts/events/#events--logs-and-filtering

```
getLogs({
  address: addr,
  fromBlock: i,
  toBlock,
  topics: [
    [
      maci.filters.Init().topics?.[0],
      maci.filters.SignUp().topics?.[0],
      ...
    ]
  ]
})
```

Reply:
Thanks for the suggestion, I'll take a look these days

Reply:
@yuetloo Would you have an example? the topics seems to be related to a specific event, it seems I cannot add topics of different events in that array. However, I did remove the many sleeps in favour of a sleep at the end of each loop iteration to be more performant.